File: src/main/java/net/golem/raknet/protocol/connected/request/OpenConnectionRequest2Packet.java
package net.golem.raknet.protocol.connected.request;

import net.golem.raknet.codec.PacketDecoder;
import net.golem.raknet.codec.PacketEncoder;
import net.golem.raknet.protocol.DataPacket;
import net.golem.raknet.protocol.RakNetPacketIds;

import java.net.InetSocketAddress;

public class OpenConnectionRequest2Packet extends DataPacket {

	public long clientId;

	public InetSocketAddress serverAddress;

	public short maximumTransferUnits;

	public OpenConnectionRequest2Packet() {
		super(RakNetPacketIds.OPEN_CONNECTION_REQUEST_2);
	}

	@Override
	public void encode(PacketEncoder encoder) {
		encoder.writeMagic();
		encoder.writeAddress(serverAddress);
		encoder.writeShort(maximumTransferUnits);
		encoder.writeLong(clientId);
	}

	@Override
	public void decode(PacketDecoder decoder) {
		decoder.skipMagic();
		serverAddress = decoder.readAddress();
		maximumTransferUnits = decoder.readShort();
		clientId = decoder.readLong();
	}

	@Override
	public String toString() {
		return "OpenConnectionRequest2Packet{" +
				"clientId=" + clientId +
				", serverAddress=" + serverAddress +
				", maximumTransferUnits=" + maximumTransferUnits +
				'}';
	}
}

File: src/main/kotlin/raknet/packet/protocol/connected/request/OpenConnectionRequest2Packet.kt
package raknet.packet.protocol.connected.request

import io.netty.buffer.ByteBuf
import raknet.Magic
import raknet.Magic.readMagic
import raknet.packet.DataPacket
import raknet.packet.PacketType
import raknet.readAddress
import java.net.InetSocketAddress

class OpenConnectionRequest2Packet(
    var magic: Magic,
    var serverAddress: InetSocketAddress,
    var mtuSize: Int,
    var clientGuid: Long,
): DataPacket(PacketType.OPEN_CONNECTION_REQUEST_2.id()) {

    override fun encodeOrder(): Array<Any> {
        return arrayOf(magic, serverAddress, mtuSize, clientGuid)
    }

    companion object {
        fun from(data: ByteBuf): OpenConnectionRequest2Packet {
            return OpenConnectionRequest2Packet(
                data.readMagic(),
                data.readAddress(),
                data.readUnsignedShort(),
                data.readLong()
            )
        }
    }

    override fun toString(): String {
        return "OpenConnectionRequest2Packet(magic=$magic, serverAddress=$serverAddress, mtuSize=$mtuSize, clientGuid=$clientGuid)"
    }

}